python/pyspark/ml/tuning.py

```diff
@@ -207,6 +210,205 @@ def _to_java_impl(self):
         return java_estimator, java_epms, java_evaluator
 
 
+class _ValidatorSharedReadWrite:
+
+    @staticmethod
+    def saveImpl(path, instance, sc, extraMetadata=None):
+        from pyspark.ml.classification import OneVsRest
+        numParamsNotJson = 0
+        jsonEstimatorParamMaps = []
+        for paramMap in instance.getEstimatorParamMaps():
+            jsonParamMap = []
+            for p, v in paramMap.items():
+                jsonParam = {'parent': p.parent, 'name': p.name}
+                if (isinstance(v, Estimator) and not (
+                        isinstance(v, _ValidatorParams) or
+                        isinstance(v, OneVsRest))
+                    ) or isinstance(v, Transformer) or \
```

Comment:
The Validators class will directly take `Estimator` and `Evaluator`, and the `Transformer` will be part of the pipeline Estimator. Should the `Transformer` params be part of the pipeline params?

Reply:
> It would be clearer if we can create an interface similar to DefaultParamsWritable

Pyspark already has this interface, the issue is the existing pyspark estimator/transformer did not inherit `DefaultParamsWritable`, change them to inherit `DefaultParamsWritable` may cause breaking changes.

Reply:
> The Validators class will directly take Estimator and Evaluator, and the Transformer will be part of the pipeline Estimator. Should the Transformer params be part of the pipeline params?

A pyspark param value can be estimator/transformer/evaluator. They're all legal.
Although currently pyspark does not have the case "transformer" to be a param value,
but, allow it here is to provide extensibility.

```
if (isinstance(v, Estimator) and not (
                        isinstance(v, _ValidatorParams) or
                        isinstance(v, OneVsRest))
                    ) or isinstance(v, Transformer) or \
                        isinstance(Evaluator):
```
The logic try to keep equivalent with jvm side logic `isInstanceOf[DefaultParamsWritable] && !isInstanceOf[MLWritable]`

Reply:
> A pyspark param value can be estimator/transformer/evaluator. They're all legal.
Although currently pyspark does not have the case "transformer" to be a param value,
but, allow it here is to provide extensibility.

Not allowing them may reduce the confusion by a lot. It's easy to add the extension when that case is explicitly supported, right?

Reply:
> It's easy to add the extension when that case is explicitly supported

Seems no.
Suppose there're some 3rd party libraries depends on spark. And 3rd party estimators may include an evaluator param (e.g. for supervising in training early stop) . We need to support them.